### PR TITLE
debaml: parity-decline map-containing dynamic parse schemas

### PR DESCRIPTION
**Pull request by invakid404**
## Problem

The de-BAML e2e differential over-claims map-containing dynamic parse schemas.
`TestBamlfuzzDynamicOracle/corpus/map_of_ints` and `corpus/map_to_class` fail
the **schema-key-ORDER** assertion on the REST leg under
`BAML_REST_USE_DEBAML=true`. The semantic `parse_diff` PASSES — native produces
a semantically-correct map — but native emits map entries in **parsed input-key
order**, which cannot be proven to equal BAML's observable map-key order under
preserve-order (map order is user-visible). Under parity-decline discipline, a
value whose ORDER native cannot prove is still an over-claim.

## Fix (decline, not match)

Add a `checkNoMap` gate so `internal/debaml.Parse` (and `parseStream`, which
shares `checkSupported`) returns `bamlutils.ErrDeBAMLParseUnsupported` for a map
in **any reachable position** — class field, list element, map value, tuple
item, or union arm, **including the non-null arm of an optional map** that
`checkSupportedType`'s nullable fast path (`if u.Nullable { return nil }`) would
otherwise wave through. The generated seam
(`codegen_debaml.go maybeParseDeBAMLFinal`) already falls back to BAML CFFI on
the sentinel, so BAML owns the authoritative map order.

`coerceMap` and the map gate (`checkSupportedMapKey`, `checkUnionMapVariant`,
`tryCastMap`) are left **unchanged** — retained but unreachable from `Parse` —
for a future PR that re-enables maps behind a proven map-order fixture. Codegen
and the render path are untouched.

## Tests

- The native-map unit tests now drive the unchanged `coerceMap` through a
  gate-bypass helper (`coerceViaBundle` / `coerceStreamViaBundle`), so map
  coercion behavior stays fully covered.
- New `TestParse_MapContainingSchemasDecline` + `TestCheckNoMap_Positions` pin
  the new gate (top-level, nested, union-arm, class-ref, optional-map, stream).
- `TestParse_NullableUnsupportedArmClaimsNull` reworked to a map-free
  unsupported arm so the nullable-fast-path property stays testable.
- The bamlfuzz parse-recovery differential forces map-containing cases to the
  fallback disposition (`parseRecoverySchemaContainsMap`) rather than flipping
  ~29 interleaved pins; the map pins remain as documentation of the coerceMap
  behavior the future PR re-enables.

## Validation

- `gofmt` / `go build ./...` / `go vet ./...` clean.
- `go test ./internal/debaml/...` green.
- `GOWORK=off go test ./codegen` (adapters/common) green.
- bamlfuzz parse-recovery differential (Docker) **green** — 54 final-parse cases
  fall back (maps now decline), no cut-line drift, native declines maps → BAML
  fallback → identical output.
- `TestBamlfuzzDynamicOracle/corpus/{map_of_ints,map_to_class}` **green** under
  `BAML_REST_USE_DEBAML=true` (the target failures).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

<a href="https://cubic.dev/pr/invakid404/baml-rest/pull/579?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

## Summary by CodeRabbit

* **Bug Fixes**
  * Map-containing schemas now consistently decline in both parse and stream paths, including nested maps and maps reached via other schema constructs.
  * Nullable and edge-case scenarios now align more predictably with the same parity-decline behavior, while coercion-based map handling remains available for the expected outcomes.
* **Tests**
  * Updated map/union expectations to validate the coercion/claim contract instead of the older parse/unsupported sentinel behavior.
  * Added coverage to verify map rejection across many reachable schema positions.
* **Documentation**
  * Refreshed the “map” section guidance for parse recovery claim vs fallback behavior.

### CODE REVIEW

File: internal/debaml/coerce_array_to_singular_test.go [hunk 1]
```diff
@@ -179,7 +179,7 @@ func TestUnion_MapArmNonStringKeyDeclines(t *testing.T) {
 		&bamlutils.DynamicTypeSpec{Type: "map", Keys: litKey, Values: &bamlutils.DynamicTypeSpec{Type: "int"}},
 		&bamlutils.DynamicTypeSpec{Type: "string"},
 	)
-	requireUnsupported(t, s, `{"u":{"a":1}}`)
+	requireCoerceUnsupported(t, s, `{"u":{"a":1}}`)
```
[the remark this answers is not in the document]
Comment (invakid404) [reply internal/debaml/coerce_array_to_singular_test.go:190]: Fixed in de1d4ef7. The comment now explicitly documents that this test uses the coerceViaBundle gate-bypass helper and exercises the coerce-time union-shape reproof path (`coerceUnionSafe` -> `checkSupportedUnionShape` -> `checkUnionMapVariant`), not the public `checkNoMap` gate.
